### DISCUSSION TITLE
Why does `docker login` try to rewrite `~/.docker/config.json` on every run?

### DISCUSSION THREAD

**Discussion by derekpovah** [moby/moby#48529] [Q&A] [answered]
I'm not sure why `docker login` has to modify `~/.docker/config.json` if it is using existing credentials. 

On NixOS, this poses a problem because `config.json` is a symlink to a write-protected file that's built by my home manager config.

```bash
$ readlink ~/.docker/config.json
/home/derek/.config/sops-nix/secrets/ghcr_auth

$ readlink -f ~/.docker/config.json
/run/user/1000/secrets.d/28/ghcr_auth
```

`docker login` fails with: 
```
Error saving credentials: rename /home/derek/.docker/config.json2180380217 /home/derek/.config/sops-nix/secrets/ghcr_auth: invalid cross-device link
```

Maybe someone with more insight can explain why this rewrite needs to happen.

**Comment by thaJeztah** [comment, marked as answer]
`docker login` (when not using a credentials-helper) stores credentials in the CLI's config file (this was a bit of a poor choice to store credentials together with other configuration in the same file); it takes the credentials provided by the user, sends them to the registry to verify if they work, and on success, it creates (or updates) the configuration file with those credentials.

Currently, that last step does not involve comparing the credentials with those that are (possibly) present in the configuration file. This probably is an optimisation step that could be made (i.e., if they're identical, then skip updating the file).

The `rename /home/derek/.docker/config.json2180380217` step is due to the use of an atomic writer; I think we introduces this to work around issues with WSL2, where updating a file in-place sometimes resulted in file corruption; the atomic writer writes to a temporary file, and then swaps the original file.

It's worth opening a ticket in https://github.com/docker/cli, which is the repository where this code is maintained; the optimisation step sounds like something that could be considered implementing.

**Reply by thaJeztah** [reply]
cc @laurazard (FYI)

**Reply by thaJeztah** [reply]
I opened a pull request to improve this;

- https://github.com/docker/cli/pull/5553

**Reply by laurazard** [reply]
I've actually been working on a larger refactor of all of this, so that's great timing :)

**Comment by aaron-bentley-imax** [comment]
It's true that this can be optimized out in the case where no updates are needed.  But if any updates *are* needed, you're back to where you started.  So either:
1. the updates are never needed, and this step can be removed
2. the updates are sometimes needed, which is again incompatible with NixOS.

In the second case, you may wish to consider creating a temporary copy of `~.docker` using `--copy-contents` and pointing DOCKER_CONFIG at it.  That should work regardless of whether updates were needed.

**Reply by thaJeztah** [reply]
The file (and directory) is expected to be writable to update config / state, so effectively the NixOS config is incompatible. The changes from https://github.com/docker/cli/pull/5553 were merely to make the CLI act nicer in a read-only situation where there's no good reason to try to update the file, but I don't think we need to go beyond that, because these files are currently expected to be writable.

**Reply by aaron-bentley-imax** [reply]
Sorry, I was addressing this to the original reporter.  I don't think there's more for Docker to do here.  In fact, this isn't even a problem with vanilla NixOS, just for people using the popular Home Manager extension to manage their docker config.

**Reply by thaJeztah** [reply]
Oh! My mistake! No harm done ❤️
